### ISSUE TITLE
Appetite for rendering streams for performance gains in the future?

### ISSUE THREAD

**Issue by dalmaer** [turbolinks/turbolinks#152]
I was curious if there would be any interest in enhancing the XHR approach that is there now to also support fetch with streamable readers as an option?

This would allow the view to not have to wait for the content before flushing it, and thus a perceived performance gain, using one of the nice features of the Web.

https://googlechrome.github.io/samples/fetch-api/fetch-response-stream.html

I know that this is very new to the platform, so there is far from a rush to deliver it, but if there was interest we may be able to get a PR together to see if there are any wins.

**Comment by nateberkopec** [comment]
This is very cool!

Since Rails can stream responses, I am 👍 for this.

[The only problem is that the fetch API isn't widely supported yet.](http://caniuse.com/#feat=fetch) There are polyfills, but I doubt they can polyfill response streaming.

**Comment by dalmaer** [comment]
Thanks Nate.

The approach I was thinking of here wasn't to try to polyfill, but rather
one of progressive enhancement.

If the browser supports service workers, install one, and it would then
handle requests and stream. This was you are guaranteed fetch() and
friends. Does that make sense?

**Comment by nateberkopec** [comment]
I don't think we need serviceworkers for this. Support for ServiceWorkers is slightly worse on Android, so if you could do it purely with `fetch` that would be better.

I'm not sure the benefit is big enough though to take a "if fetch do this else XHR" approach here. It's more of a "nice-to-have" than "must-have", at least for Turbolinks core.

Looking at making this as some kind of plugin, perhaps Turbolinks could support `event.preventDefault()` in `turbolinks:request-start`?

**Comment by nateberkopec** [comment]
Plus, Rails streaming [is kinda janky anyway](https://github.com/rails/rails/issues/23828).

**Comment by javan** [comment]
It's worth noting that `fetch` doesn't currently provide a way to abort requests, something that [Turbolinks supports via XHR](https://github.com/turbolinks/turbolinks/blob/c73e134731ad12b2ee987080f4c905aaacdebba1/src/turbolinks/http_request.coffee#L20-L22). Long discussion about it here: https://github.com/whatwg/fetch/issues/27

**Comment by nateberkopec** [comment]
Bummer. I wonder if it would really break anything though if currentVisit.cancel() was a no-op.

It looks like you can't timeout a fetch either really. You can settimeout and then ignore the result of the fetch, but it's still there.

**Comment by dalmaer** [comment]
The benefit of SW was that we could support offline use cases better, and
could even setup races for content etc. That felt like the biggest benefit
not fetch vs xhr.

**Comment by nateberkopec** [comment]
I guess you can stream with XHR anyway, at least in a Turbolinks context (assuming people aren't streaming like 10MB+ response bodies...).

I've often thought ServiceWorkers would be a pretty natural extension to a Turbolinks approach. Not sure where they would plugin either.

**Comment by jpic** [comment]
Definitely agree with @nateberkopec that providing a serviceworker.js will be a major step towards making turbolinks even moar awesome. I tried it yesterday and after a few hours I had something functional for a frontend that didn't have turbolinks. The quality of the JS API around cache is surprisingly clean and efficient.
